bisection.py

```diff
@@ -0,0 +1,373 @@
+            return False
+        left_commit = self.torch_src.commits[0]
+        right_commit = self.torch_src.commits[-1]
+        self.bisectq.append((left_commit, right_commit))
+        return True
+        
+    def run(self):
+        while not len(self.bisectq) == 0:
+            (left, right) = self.bisectq.pop(0)
+            left.score = self.bench.get_score(left)
+            right.score = self.bench.get_score(right)
+            if self.regression(left, right):
+                mid = self.torch_src.get_mid_commit(left, right)
+                if mid == None:
+                    self.result.append((left, right))
+                else:
+                    mid.score = self.bench.get_score(mid_commit)
+                    if self.regression(left, mid):
+                        self.bisectq.append(left, mid)
+                    if self.regression(mid, right):
+                        self.bisectq.append(right, mid)
+ 
+    def output(self):
+        json_obj = dict()
+        json_obj["start"] = self.start
+        json_obj["end"] = self.end
+        json_obj["threshold"] = self.threshold
+        json_obj["timeout"] = self.bench.timelimit
+        json_obj["torchbench_branch"] = self.bench.branch
+        json_obj["result"] = []
+        for res in self.result:
+            r = dict()
+            r["commit1"] = res[0].sha
+            r["commit1_time"] = res[0].ctime
+            r["commit1_score"] = res[0].score
+            r["commit2"] = res[1].sha
+            r["commit2_time"] = res[1].ctime
+            r["commit2_score"] = res[1].score
+            json_obj["result"].append(r)
+        with open(self.output_json, 'w') as outfile:
+            json.dump(json_obj, outfile, indent=2)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument("--work-dir",
+                        help="bisection working directory",
+                        type=exist_dir_path,
+                        required=True)
+    parser.add_argument("--pytorch-src",
+                        help="the directory of pytorch source code git repository",
+                        type=exist_dir_path,
+                        required=True)
+    parser.add_argument("--torchbench-src",
+                        help="the directory of torchbench source code git repository",
+                        type=exist_dir_path,
+                        required=True)
+    parser.add_argument("--start",
+                        help="7-digit SHA hash of the start commit to bisect",
+                        required=True)
+    parser.add_argument("--end",
+                        help="7-digit SHA hash of the end commit to bisect",
+                        required=True)
+    parser.add_argument("--threshold",
+                        help="the torchbench score threshold to report a regression",
+                        type=float,
+                        required=True)
+    parser.add_argument("--bmfilter",
```

Comment:
what happens if unspecified?

i think it's good to have a mode that focuses on one test like we discussed.

but when running a 'score' driven bisection, specifying filter is a little awkward.  instead might be easier to just specify score version and the filter is derived from the score config?

Reply:
Fixed by using bisection config yaml.